### ISSUE TITLE
Determine reflection MVP

### ISSUE THREAD

**Issue by seanmiddleditch** [potatoengine/potato#97]
Determine what kind of system we want for reflection.

Some options:
- C++ reflection (like Prior Engine X)
- Swig-like C++ parsing (maybe with clang)
- DDL and code generation

Criteria:
- Minimal runtime overhead
- Easy to use
- Helps with language binding
- Helps with editor binding
- Helps with remote control binding
- Engine data serialization

**Comment by marcinwie** [comment]
since other use cases are listed -- i would like to throw in engine data serialization as criteria.

**Comment by seanmiddleditch** [comment]
The following is a bunch of thoughts about reflection implementation. I'll make a few of these comments as I lay out my thoughts and to solicit feedback.

**Pure C++ Reflection**

Prior Engine X used a completely C++ mechanism that implied a decent chunk of run-time overhead and syntactical shenanigans. Reflection looked something vaguely like:

```c++
struct Type {
  int x, y;
  int sum() const { return x * y; }
};

REFLECT(Type)
  [reflect::readonly]
  [reflect::help("Horizontal component")]
  .field(&Type::x)

  [reflect::help("Vertical component")]
  [reflect::range(0, 1000)]
  .field(&Type::y)

  [reflect::expose]
  .method(&Type::sum);
```

The actual code was designed back for C++03 and was a bit uglier than the above, but close. I think with C++17 and the magic of `constexpr` I can make that code as clean or cleaner than the above sample and potentially even get rid of most or all of the run-time initialization overhead.

However, there was another downside to that system: its run-time implementation at _best_ could be described as "meh" (and a `constexpr` version would likely be even worse). The main problem was run-time performance at point of use.

If one wanted to, say, serialize the data, they'd have to iterate the fields one by one, invoke a getter method, and then pass that through a visitor class. To avoid the overhead of something like `any` this visitor class only worked on primitive types, so a member structure or field containing a complex object required a recursive visitor (so an object would be serialized by calls to `enterField(name)`, `beginObject`, ...fields..., `endObject`, `leaveField`; each individual field would be serialized with calls like `enterField(name)`, `writePrimitiveTypeHere(value)`, `leaveField`).

To put it simply, the speed of serialization was _atrocious_. Since all the complexity of handling non-primitive types lived in serialization (to avoid `any` types) this meant that *all* interaction with reflection occurred through serialization. More on that later.

While I think we can do better in terms of serialization overhead, I don't think we can do _much_ better... some type erasure can be applied, but it will ultimately always boil down to virtual function calls and primitive values, and _maybe_ some abstract interfaces to wrap collections and like (see below).

Collections. So, this serialization system was just primitives. It didn't really have a concept of arrays or containers. That ultimately meant that a `vector<Type>` might be serialized with calls like `beginField`, `enterObject`, `beginField("count")`, `writeInt`, `endField`, ... `beginField` `writeFoo` `endField`, `leaveObject`, `endField`. Note that nothing really says "this is a list of things." There's just a field that _by convention_ has the name `"count"` which included the list of items, but the reading end really didn't know what that meant. And because serialization was the only way to interact with reflection, this is what UI bindings in the editor had to use. The editor has no idea what was just a list of items and what was a resizable `vector` (or other container) of items. This meant that basic UI like "add a new item" or "delete the second item" or whatnot was not present! I would non-hyperbolically claim that the lack of such a basic element of editing data cost us man-years of time/money on that project across everyone who had to interact with those tools.

The reflection system _did_ allow for custom attributes which could be any C++ type, so we eventually added some attributes that exposed `resize`, `remove`, `insert`, etc. operations, but they were not automatically applied to `vector` or the like (that was hard to do; I think we can do that a lot better today) and due to C++'s type system they were cumbersome to specify (again, I think we could do better there with C++17). That made the problem less severe in the end but it was still a far cry from ideal.

As a final note, while some reflection systems embed more of this magic into headers and templates, we want to avoid that: template overhead is a _bug_ that we try to avoid like the plague. Imposing _significant_ compilation overhead into effectively every TU for any type that they might use is entirely unacceptable. It's critical that all the heavy lifting of reflection remain in TUs, which in turn imposes that some level of access remains via `virtual` functions.

The major short-comings of this approach then were clunkiness (we could fix that) and performance (we could only slightly improve that).

**Actionable Notes**

If we decided we wanted to follow this approach, I'd advise a few changes for fellow engineers familiar with Prior Engine X.

For one, the reflection builder should be `constexpr` templates. Appending fields and methods should not require run-time overhead nor memory allocation. We can build the reflection type iteratively with signatures like: `constexpr auto builder<T>::field(T::M* ptr) const noexcept { return builder<decltype(*this), field(decltype(ptr), ptr)>{};` (mostly gook, but the idea is that we return a new builder type for each `.field` call that has the new field encoded _in the type_).

For two, attributes should also be defined as templates, so that they can mutate the builder state (the same way as adding a field) and hence attributes can actually know the real type of the things they're modifying.

However, while both of these will cut down on overhead and esp. initialization cost, they won't help with run-time lookup costs, and they _will_ impose heavier compile-time overhead to any TUs that actually define reflection data.

**Preliminary Conclusion**

We could do worse than using this system (with outlined modernization/improvements), and that due to our collected experience such a system would be relatively easy to implement and easy (for us) to use. But, I think that _run-time lookup_ overhead is going to be killer and that we can do a _lot_ better than that, albeit at greater (initial) development cost and that we _should_ strive to do a lot better.

**Comment by seanmiddleditch** [comment]
More rumination.

**TDF and Schema Generators**

Tools like [Google FlatBuffers](https://github.com/google/flatbuffers) rely on a custom schema file and a custom code generator to convert DDL (data description language) into in-memory data structures for C++ (and other languages) as well as generating efficient serialization code.

In practice, this works _really_ well. A relatively large portion of Internet traffic is now encoded in these kinds of system (e.g. [ProtoBuf](https://developers.google.com/protocol-buffers/), [Apache Thrift](http://thrift.apache.org/), etc.).

Those system work with a schema that is meant to be _version independent_, but did that via convention: a user was never supposed to remove or change an existing field, and was never supposed to add a new non-optional field, so that old data continued to work. A schema is (roughly) something like:

```
type foo {
  x : int = 1,
  y : int = 2,
  z : int = 3 [optional]
}
```

A tool would read that schema and generate C++, Java, Python, whatever that would have an appropriate data-type and reader/writer methods that efficiently streamed bytes.

TDF was a format (I think it stood for Table Data Format or something like that) built back at GPG that described a _changelog_ of a data format. Instead of a schema, it looked something like (in C#, but here in pseudo-code):

```
type foo {
  version 1 {
    x : int
    y : int
  }
  version 2 {
    add z : int
  }
}
```

Just like the other schema generates, it would generate an in-memory data structure for the final version, and read/write methods. However, the tool could also run in a conversion mode, using specified default values or even expressions (as C# closures) that would run as part of an update process. This allowed data formats to change in non-trivial ways but allowed old data to seamlessly (usually) be converted to still work.

The downside of course is that downgrading wasn't really a thing; converting newer data to run on older code was out of the question. However, it generally worked a lot better _in development_ than tech like ProtoBuf, which assumes that a protocol once published could never break (which isn't a good choice for data-storage for a multi-year-long game development project).

**Limitations of Code Generation**

All of the above approaches generate relatively simple in-memory data structures. One would have trouble generating a "rich class" in a language like C++ using those techniques. C# or Python makes this easier, but we mostly care about C++ for Potato.

The approach necesary then is ot generate "data classes" and embed those as fields in a richer class, something like:

```c++
/// generated
struct FooData {
  int x, y, z;
};

/// manual wrapper
class Foo {
public:
  int x() { return _data.x; }
  ...
private:
  FooData _data;
};
```

That's boilerplate-heavy, not really inline with modern C++ design practices (don't write unnecessary wrappers!), and most workarounds have their own limitations (e.g. inheritance could work, but you can't then encapsulate just some fields, and object slicing becomes a problem).

However, it does mesh well with a pure-data ECS approach, since we'd just want simple structs for Components anyway. That doesn't really help with all the _other_ places reflection/serialization might be useful, though.

**Hypothetical Serialization Superpowers**

The nice thing about code generation for read/write is that the generated code is _fast_. There's no `virtual` methods that don't need to be there. It can do raw `memcpy` where appropriate and use simple pointer offsets and the like. It can be maximally-efficient code generated and then compiled right by our cutting-edge C++ compilers.

Hypothetically, though, there's no reason we should be limited to only read/write. Or one version of read/write. It's entirely possible that we use a text-template engine or similar tech to allow generation file I/O serialization functions and separate network serialization functions, and separate script binding functions, and separate UI/property-grid functions. We can generate any permutation of code necessary to maximally-efficient serialization in each case, concerned only with the binary code bloat if we overdo it.

Note that, unlike the run-time based serialization systems out there, this generated code is _fully debuggable_. A dev can set a breakpoint in a specific type. They can edit code. They can develop just like it's any other code, which is a huge boon.

**Multi-language Support**

We're currently discussing Python tools scripting and other scripting language support. Using a schema-driven data definition allows us to generate data structures and code for more than just C++, which could be a huge interoperability help. Marshalling data through C++ can be a pain when you're just trying to write a simple stand-alone Python utility, for instance.

**Attributes**

Keeping attributes in schema files has its pros and cons. It potentially allows them to be defined in a more natural syntax than the best we can do in C++ today, and with less overall code in the engine.

These attributes would then be used by the code generators and potentially never even be written out or exposed in the generated code or data, which ultimately can mean less code and smaller data files. Hypothetically.

**Downsides**

Supporting richer data types in schema files imposes a cost, both on the schema file parser and on _all_ code generators. There's also potentially interop concerns; e.g. if a dictionary type exists, we want to be careful that round-tripping data through any language doesn't alter the output (e.g., doesn't re-order keys), but this can be painful when every language has its own optimal dictionary.

Another downside of course is the initial development cost. The system I've described here is only tangentially similar to ProtoBuf or the like, so it would almost certainly need to be developed from scratch. Even developing it in Python (which would be my inclination), it's not a trivial task.

The absolute biggest downside is just needing to even have a bespoke DDL used for all data definition in the engine. This creates a barrier to entry and contextual overhead to other developers using the engine. It forces a data away from code that operates on it, which can hinder code readability.

**Preliminary Conclusion**

I think that _in the end_ this solution is strong, primarily because of its ability to generate maximally-efficient serialization routines. However, the cost to initial development is high, and the continued maintenance burden is higher than the pure C++ approach.

It also imposes burden into the build system, which would have to understand when and how to regenerate code as schemas change. If we do the TDF-style approach of allowing changes to schemas, we also need a process to auto-convert data when a user gets newer code or a newer binary but has old data.

**Comment by seanmiddleditch** [comment]
Another braindump on a potential avenue.

**C++ Parser for Reflection**

Some engines make use of code generators that parse regular or specially-annotated C++ code to generate reflection metadata. [Unreal does this via its UProperty system](https://wiki.unrealengine.com/UPROPERTY), for example. [Swig](http://www.swig.org/) has been used for this purpose for scripting bindings for years. Some proprietary AAA engines reportedly use mechanisms like this (none with public documentation that I'm aware of). There's a number of open source repos like [clReflect](https://github.com/Celtoys/clReflect) that do things with a Clang-based parser.

The idea is that code and data structures are written in plain ol' C++ with a minimum of fuss, but the parser+generator allows to gain many or most of the advantages of custom high-performance serializers and other specialized bindings.

**Compatibility**

The biggest cost of such an approach is compatibility. Many of these tools are incapable of parsing the entirely of C++03, much less the latest version of C++ that Potato targets.

Clang-based tools certainly have more flexibility here, of course, but it's not guaranteed to be a drop-in replacement and requires constant maintenance (e.g., updating the Clang version used, which does not maintain a stable API/ABI).

**Performance**

The second biggest cost of this approach is the performance of compilation. Changes to affected headers (including their included headers, potentially) requires re-running this tool. Simpler parsers might be able to parse the C++ relatively quickly, but then they incur a higher incompatibility cost. Highly-compatible parsers like those based on Clang effectively require running the compiler frontend twice, which is naturally more expensive than just running it the once for regular compilation.

**Preliminary Conclusion**

I think this might be the best way to implement reflection in general (sans getting proper support in the language), but the maintenance cost is high enough that it would need to be a bigger open source project than something used by just Potato, and the existing open source solutions out there are either unmaintained or insufficiently complete for our purposes.

**Comment by seanmiddleditch** [comment]
And one more.

**Native Reflection in C++ 2?**

Reflection is incoming for standard C++. The whole [ISO WG21 SG7 - Reflection](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3814.html) study group is working away on that, and there's already a [Static Reflection TS](https://github.com/cplusplus/reflection-ts) out for IB vote. (That proposal is not the syntax that C++ reflection will take; the paper is being evaluated for functionality, not syntax.)

The reflection currently under consideration is _static reflection_. That is, it's the ability to write code that reflects over C++ during compilation. I believe that such a mechanism is more than enough to write any dynamic reflection necessary, and odds are good that enough code generation support will be present to write (reasonably) efficient custom serializations (reasonably) easily.

The biggest problem is that since all this reflections runs in the C++ compiler, there's going to be a lot of gymnastics to keep it fast. Complicated reflection will need to remain in TUs and will need to be acceptbly fast that we can stomach recompiling those TUs whenever a header (or module interface) changes.

I believe ultimately that just exporting the reflection information and letting an external tool and the build system run code generation is a superior approach for _large_ reflection problems, but the native C++ reflection is going to be immensely useful and powerful for _small_ reflection problems.

**The Immediate Problem**

However, realistically, this is a ways out. C++20 isn't even out yet. Reflection will _at the earliest_ be in C++23, but I wouldn't be at all surprised if it is held back until C++26 or later.

In short, it might be the best thing ever, but it is of absolutely zero use to use for almost the next decade at the least.

**Final Conclusion**

The Potato of the future may be making excellent use of standard C++ reflection, but the Potato today simply doesn't have that option. Native C++ reflection is not an option right now.

**Comment by seanmiddleditch** [comment]
**Hybrid Solutions**

There's a few solutions we've hypothesized about using on previous projects that might work.

**Native C++ Reflection without Schema Generation**

This is essentially the reverse of using a schema to generate code, sort of. Essentially, we write regular C++ code and use native C++ libraries for basic usage. A particular tool links against the C++ code and walks all the reflected types, generating and writing out a schema for each.

That schema can then be used for all the usual purposes, like generating efficient serialization code or for interop with other languages.

The benefits are that the data types can live in real C++ and be full classes and use arbitrary types and all that. The downsides are that this basically requires all the work of both native C++ and schemas, plus some.

**Swig Schema Generation**

Same as above, but using Swig or a similar C++ parsing tool to extract the schemas. This hypothetically has all the advantages of using Swig to generate the serialization directly, but gives the benefits of schema intermediaries (like generating multiple language bindings or the like).

**Linker/PDB Shenanigans**

A few engines have used techniques where they parse object files, DLLs, and PDBs (or DWARF equivalent) to find out about data structures and the like.

This is similar to using a tool like Clang/Swig to parse the C++ source, but has the added benefit of both being a full C++ parser (because it's your regular compiler) and of avoiding extra C++ overhead (because you're only compiling each TU once still).

The downside is that it can be difficult to extract all desired meta-information this way, and it's highly platform-specific. e.g. an ideal implementation would work with both PDBs for Windows and with DWARF debug info for Linux/macOS, and that still wouldn't be comprehensive (though it would be in practice; there's no platform we care about that isn't based on either of those categories of toolchain).

**Manual Serialization**

This is the low-tech option. Basically, don't do reflection with any automatic means. Require users to write functions that do serialization manually.

This is repetitive and _can_ be error-prone (depending on what those manual interfaces look like). It does allow hyper-optimization when needed, potentially even more so than any potential automatic version.

As an example of a simple version, the basics of reflection might just require a function like this for every reflected type:

```template<typename Serializer>
auto serialize(MyType& val, Serializer& ser) {
  ser.field(val, &val->field);
  ser.field(val, &val->other_field);
  ser.function(val, &MyType::method);
}
```

One advantage over other native C++ techniques is that there's no need to wrap fields/functions and store them off in type-erased data structures. Another advantage is that the functions can hypothetically be overloaded and the like for specialized/optimized cases. The lack of type-erasure makes it a lot easier for the optimizer to do a good job.

It's arguable how much worse this is than other native C++ reflection libraries. With the right care in interface, the amount of typing and error-prone-ness for this vs type reflection is pretty similar.

The biggest problem is that - assuming the template nature - is that this has to live in a header. So long as the interface is kept thin, that shouldn't impose a lot of cost on compilation, but it's certainly not free.

A particular problem to keep in mind is `const`-correctness. To allow a single serialize function to operate in both read and write mode (necessary to avoid redundancy and becoming error-prone), the `serialize` function needs to be specified in some way that works with both lvalue and const-lvalue values. The obvious solution removes the ability to overload easily on the type without a lot of obnoxious SFINAE work.

**Comment by seanmiddleditch** [comment]
Some off-the-shelf technologies to consider:

**Schema Based**

- [FlatBuffers](https://google.github.io/flatbuffers/flatbuffers_guide_tutorial.html)
- [CapnProto](https://capnproto.org/)
- [MessagePack](https://github.com/msgpack/msgpack-c)

**C++ Reflection/Serialization**

- [RTTR](https://www.rttr.org/)
- [cereal](https://github.com/USCiLab/cereal)
- [cppbor](https://github.com/rantydave/cppbor)
- [Helium Reflect](https://github.com/HeliumProject/Core/blob/master/Documentation/Module-3-Reflect.md)
- [meta](https://github.com/skypjack/meta)
- [ponder](https://github.com/billyquith/ponder)

**Clang Reflection**

- [clRefect](https://github.com/Celtoys/clReflect)
- [cppast](https://github.com/foonathan/cppast)

**Comment by seanmiddleditch** [comment]
So I played around a bit with writing a modern C++17 native-C++ reflection library with the ideas I outlined above. It's all feasible, but... what a slog. I don't think it's a worthwhile endeavor with "real" reflection in C++ right around the corner, and with all the open source libraries linked above (though I somewhat dislike each and every one of them, for varying reasons).

I'm leaning a little bit towards the template `serialize` function I outlined a few comments ago. It's not perfect, *but* it has a few hypothetical strengths:

- It's relatively easy to develop.
- It allows reasonable optimizations.
- It's easy to transition to real C++ reflection down the road.
- It's easy to write helpers/macros to make it even easier to write/use.
- Those same macros make it easy to write a Python script that extracts Reflection if we want a schema down the line.

I'm going to take a crack at building that out and see what it looks like before making a call.

**Comment by seanmiddleditch** [comment]
I started on the `serialize` function implementation. For the most part, it works as I thought it would. Downsides are that writing serializers/reflectors that hit maximum performance is non-trivial, and a number of helpers are going to be needed to make it feasible for non-C++-nerds to get into.

I also took a deeper look at Flatbuffers. Integrating it was a little bit of a chore because the CMake build system it uses is dated and doesn't behave well when used via `add_subdirectory` in a modern CMake project; I ended up writing a new (very simplified) CMakeLists just for our purposes and using that instead of the bundled copy. After that, I took a deeper look at the generated C++. Basically, it's really built for being an interchange for manipulating data about to be persisted or just having been un-persisted, not for storage as the in-memory format. The generators are also all hard-coded into the flatc compiler, which means that extending it to other forms of serialization effectively requires re-implementing it.

In that sense, a custom schema generation system is probably going to suit our needs better than flatbuffers, though it may behoove us to use the same schema format. Assuming we end up going the schema route at all.

**Comment by seanmiddleditch** [comment]
Took a crack at using Flatbuffers.

Integration was a bit harder than desired since it uses CMake 2.8 and in ways that are antagonistic to inclusion via `add_subdirectory`. Ended up having to reimplement a simple version of its `CMakeLists.txt` to build it cleanly.

It is working and works fine. I have concerns about its design:

- The C++ code it generates is very clearly built for I/O and no other purposes
- The generators are hard-coded into the C++ binary

Given that Flatbuffers is pretty much state of the art for that style of tool, I'm thinking it's not the right approach for this purpose.

We want either to take an IDL and generate plain C/C++ code for headers and use templates to generate various reflection/serialization functions, or we want to parse C++ code and use templates.

Either way, that implies that either custom IDL engine or something like Swig will be a better choice that the Flatbuffers library.

**NOTE**: that doesn't mean not using Flatbuffers for over-the-wire or storage purposes! The *format* is fine, it's just the upstream C++ library and `flatc` tool that I don't think work for us.
